crates/codegen/parser/src/chumsky/combinator_node.rs

```diff
@@ -128,13 +135,37 @@ impl<'context> CombinatorNode<'context> {
         }
 
         tree.context.alloc_node(match &expression.ebnf {
-            EBNF::Choice(exprs) => Self::Choice {
-                name,
-                elements: exprs
+            EBNF::Choice(exprs) => {
+                enum TN<'c> {
+                    Trie(TerminalTrie),
+                    Node(&'c CombinatorNode<'c>),
+                }
+                let elements = exprs
                     .iter()
-                    .map(|expr| Self::new(tree, expr, None))
-                    .collect(),
-            },
+                    .map(|expr| {
+                        match TerminalTrie::new(
+                            tree,
+                            expr,
+                            tree.production.kind == ProductionKind::Token,
+                        ) {
+                            None => TN::Node(Self::new(tree, expr, None)),
+                            Some(trie) => TN::Trie(trie),
+                        }
+                    })
+                    .coalesce(|prev, curr| match (prev, curr) {
+                        (TN::Trie(prev), TN::Trie(curr)) => Ok(TN::Trie(prev.merged_with(curr))),
+                        pc => Err(pc),
```

Comment:
QQ: what does `pc` mean in this case?

Reply:
prev+curr. In my current PR it is renaimed to `pair`, because `pc` is obviously lazy naming.